src/core/ir/discriminator.rs

```diff
@@ -0,0 +1,1054 @@
+use std::collections::HashSet;
+use std::fmt::Write;
+use std::ops::{Deref, DerefMut};
+
+use anyhow::{bail, Result};
+use async_graphql::Value;
+use indenter::indented;
+use indexmap::IndexMap;
+
+use crate::core::config::Type;
+use crate::core::valid::{Cause, Valid, Validator};
+
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub enum TypeName {
+    Single(String),
+    Vec(Vec<String>),
+}
+
+#[derive(Clone)]
+pub struct Discriminator {
+    types: Vec<String>,
+    fields_info: IndexMap<String, FieldInfo>,
+}
+
+impl std::fmt::Debug for Discriminator {
```

Comment:
Do we need the debugs?

Reply:
I'm suggesting to leave the debug fmt because it'll help in debugging since it outputs the info stored in `FieldInfo` as type name strings instead of raw usize. I've also added `tracing::debug!` in Discriminator in case we need to debug something in runtime